File: patches/setup-wizard.py
#!/usr/bin/env python3
from pathlib import Path

p = Path('/opt/radiobot/backend/src/index.ts')
s = p.read_text(encoding='utf-8')

if "const SETUP_TOKEN = process.env.SETUP_TOKEN ?? '';" not in s:
    marker = "const DISCORD_CONTROL_ROLE = process.env.DISCORD_CONTROL_ROLE ?? '';"
    if marker in s:
        s = s.replace(marker, marker + "\nconst SETUP_TOKEN = process.env.SETUP_TOKEN ?? '';", 1)

old = "app.addHook('preHandler', async (req, reply) => { if (req.url.startsWith('/api/') && !req.url.startsWith('/api/spotify/callback') && !auth(req, reply)) return reply; });"
new = "app.addHook('preHandler', async (req, reply) => { const openSetup = req.url === '/api/setup/status' || req.url === '/api/setup/user'; if (req.url === '/api/setup' && !WEB_PASSWORD) return reply.code(403).send('Bitte zuerst einen Web-Benutzer anlegen.'); if (req.url.startsWith('/api/') && !openSetup && !req.url.startsWith('/api/spotify/callback') && !auth(req, reply)) return reply; });"
if old in s:
    s = s.replace(old, new, 1)
elif new not in s:
    raise SystemExit('API auth hook marker missing')

insert = r'''app.post<{ Body: { setupToken?: string; webUser?: string; webPassword?: string; } }>('/api/setup/user', async (req, reply) => {
  const b = req.body ?? {};
  if (WEB_PASSWORD) return reply.code(409).send('Der Web-Benutzer wurde bereits angelegt.');
  if (!SETUP_TOKEN || b.setupToken !== SETUP_TOKEN) return reply.code(403).send('Ungültiger Ersteinrichtungs-Code.');
  const username = b.webUser?.trim() || '';
  const password = b.webPassword ?? '';
  if (!username) return reply.code(400).send('Benutzername ist erforderlich.');
  if (username.length > 64) return reply.code(400).send('Benutzername ist zu lang.');
  if (password.length < 12) return reply.code(400).send('Das Web-Passwort muss mindestens 12 Zeichen haben.');
  const payload = JSON.stringify({ discordToken: process.env.DISCORD_TOKEN ?? '', webUser: username, webPassword: password, spotifyClientId: SPOTIFY_CLIENT_ID, spotifyClientSecret: SPOTIFY_CLIENT_SECRET, spotifyRedirectUri: SPOTIFY_REDIRECT_URI, youtubeApiKey: process.env.YOUTUBE_API_KEY ?? '', discordControlRole: DISCORD_CONTROL_ROLE, publicUrl: '', port: PORT, setupToken: SETUP_TOKEN });
  await privilegedConfigWrite(payload);
  return { ok: true, message: 'Benutzer erstellt. Der Server startet jetzt neu. Danach meldest du dich mit diesem Benutzer an und schließt die Ersteinrichtung ab.' };
});
'''
if "app.post('/api/setup/user'" not in s:
    anchor = "app.get('/api/setup/status'"
    if anchor not in s:
        anchor = "app.get('/api/health'"
    if anchor not in s:
        raise SystemExit('setup insertion anchor missing')
    s = s.replace(anchor, insert + "\n" + anchor, 1)

if "app.post('/api/setup/user'" not in s:
    raise SystemExit('first-user route was not inserted')

p.write_text(s, encoding='utf-8')
print('setup wizard patch applied: first web user required before setup')

File: frontend/setup-wizard.js
(() => {
  const q = new URLSearchParams(location.search);
  const hash = new URLSearchParams(location.hash.replace(/^#/, ''));
  const setupCode = q.get('setup') || hash.get('setup') || '';
  const esc = v => String(v ?? '').replace(/[&<>"']/g, c => ({'&':'&amp;','<':'&lt;','>':'&gt;','"':'&quot;',"'":'&#39;'}[c]));
  const api = async (url, options = {}) => { const r = await fetch(url, { headers: {'Content-Type':'application/json', ...(options.headers || {})}, ...options }); if (!r.ok) throw new Error(await r.text()); return r.status === 204 ? null : r.json(); };
  const style = document.createElement('style');
  style.textContent = `.setup-overlay{position:fixed;inset:0;background:rgba(2,6,23,.82);backdrop-filter:blur(12px);z-index:9999;display:grid;place-items:center;padding:20px}.setup-card{width:min(760px,100%);max-height:92vh;overflow:auto;background:#0f172a;border:1px solid #334155;border-radius:22px;padding:28px;box-shadow:0 24px 80px rgba(0,0,0,.45)}.setup-card h1{margin:0 0 8px}.setup-card p{color:#94a3b8}.setup-grid{display:grid;grid-template-columns:1fr 1fr;gap:14px}.setup-card label{display:grid;gap:7px;color:#cbd5e1;font-size:13px}.setup-card input{width:100%;box-sizing:border-box;padding:12px;border-radius:10px;border:1px solid #334155;background:#020617;color:#f8fafc}.setup-wide{grid-column:1/-1}.setup-actions{display:flex;justify-content:flex-end;gap:10px;margin-top:20px}.setup-card button{padding:12px 18px;border:0;border-radius:10px;cursor:pointer}.setup-primary{background:#38bdf8;color:#082f49}.setup-note{font-size:12px;color:#64748b}@media(max-width:700px){.setup-grid{grid-template-columns:1fr}}`;
  document.head.appendChild(style);
  function open(settings = false) {
    const old = document.querySelector('.setup-overlay'); if (old) old.remove();
    const overlay = document.createElement('div'); overlay.className='setup-overlay';
    overlay.innerHTML = `<div class="setup-card"><div class="eyebrow">MusikBot187 · ${settings ? 'Einstellungen' : 'Ersteinrichtung'}</div><h1>${settings ? 'Bot konfigurieren' : 'Willkommen bei MusikBot187'}</h1><p>${settings ? 'Alle wichtigen Zugangsdaten können hier geändert werden. Nach dem Speichern startet der Bot automatisch neu.' : 'Du musst auf Linux nur installieren. Discord, Spotify, YouTube und Web-Zugang werden hier eingerichtet.'}</p><form class="setup-grid">
      ${settings ? '' : '<label class="setup-wide">Ersteinrichtungs-Code<input name="setupToken" value="'+esc(setupCode)+'" readonly></label>'}
      <label>Discord Bot Token<input name="discordToken" type="password" autocomplete="off" ${settings ? 'placeholder="leer lassen = unverändert"' : 'required'}></label>
      <label>Web-Benutzer<input name="webUser" value="admin" autocomplete="username"></label>
      <label>Web-Passwort<input name="webPassword" type="password" minlength="12" autocomplete="new-password" placeholder="mindestens 12 Zeichen"></label>
      <label>Discord Control Role ID<input name="discordControlRole" placeholder="optional"></label>
      <label>Port<input name="port" type="number" min="1" max="65535" value="3000"></label>
      <label class="setup-wide">Öffentliche URL<input name="publicUrl" placeholder="https://musik.example.de"></label>
      <label>Spotify Client ID<input name="spotifyClientId" autocomplete="off"></label>
      <label>Spotify Client Secret<input name="spotifyClientSecret" type="password" autocomplete="off"></label>
      <label>Spotify Redirect URI<input name="spotifyRedirectUri" placeholder="https://deine-domain/api/spotify/callback"></label>
      <label>YouTube API Key<input name="youtubeApiKey" type="password" autocomplete="off"></label>
      <div class="setup-wide setup-note">Discord Token und API-Schlüssel werden mit restriktiven Dateirechten in /etc/radiobot/radiobot.env abgelegt. Für einen öffentlichen Zugang HTTPS verwenden.</div>
      <div class="setup-wide setup-actions"><button type="button" class="setup-cancel">Abbrechen</button><button class="setup-primary">${settings ? 'Speichern & Neustart' : 'Einrichtung abschließen'}</button></div>
    </form></div>`;
    document.body.appendChild(overlay);
    const form = overlay.querySelector('form');
    if (settings) api('/api/settings').then(s => { for (const [k,v] of Object.entries(s)) { const el=form.elements[k]; if(el && typeof v === 'string') el.value=v; } }).catch(()=>{});
    overlay.querySelector('.setup-cancel').onclick = () => overlay.remove();
    form.onsubmit = async e => { e.preventDefault(); const data=Object.fromEntries(new FormData(form).entries()); data.port=Number(data.port||3000); try { const r=await api(settings?'/api/settings':'/api/setup',{method:'POST',body:JSON.stringify(data)}); form.innerHTML=`<div class="setup-wide"><h2>Gespeichert</h2><p>${esc(r.message)}</p><p>Bitte einige Sekunden warten und die Seite neu laden.</p></div>`; setTimeout(()=>location.href='/',9000); } catch(err) { alert(err.message); } };
  }
  api('/api/setup/status').then(s => {
    if (s.requiresSetup && setupCode) open(false);
    const header = document.querySelector('header .row');
    if (header && !document.getElementById('settings')) { const b=document.createElement('button'); b.id='settings'; b.textContent='Einstellungen'; b.onclick=()=>open(true); header.prepend(b); }
  }).catch(()=>{});
})();
